README.md
# Puppeteer screenshot CLI

Simple wrapper around [Puppeteer](https://github.com/GoogleChrome/puppeteer) to take screenshot from command line.

## Usage

```shell
npm i [-g] puppeteer-screenshot-cli

puppeteer-screenshot --url 'http://perdu.com' --selector 'h1' --output ./perdu.jpg
puppeteer-screenshot -u 'http://perdu.com' -s 'body' -o - > /tmp/perdu.jpg
puppeteer-screenshot 'http://perdu.com' > perdu.jpg

```

### Options

```
  Headless screenshot with Puppeteer

    -u, --url string        URL to navigate page to. The url should include scheme, e.g. https://.

    -o, --output string     The file path to save the image to. If path is a relative path, then it is
                            resolved relative to current working directory. If no path is provided, the
                            image won't be saved to the disk.

    -s, --selector string   A CSS selector of an element to wait for.
                            Default: body

    -t, --type string       Specify screenshot type, can be either jpeg or png.
                            Default: png

    -q, --quality number    The quality of the image, between 0-100. Not applicable to png images.

    -w, --width number      Viewport width in pixels
                            Default: 800

    -h, --height number     Viewport height in pixels
                            Default: 600

    --timeout  number       Maximum time to wait for in milliseconds.
                            Default: 30000

    -f, --fullPage          When true, takes a screenshot of the full scrollable page.
                            Defaults: false.

    --headless              Whether to run browser in headless mode.
                            Default: true

    -?, --help              This help
```

## See also

- [puppeteer-trace-cli](https://www.npmjs.com/package/puppeteer-trace-cli)  
  Simple wrapper around [Puppeteer](https://github.com/GoogleChrome/puppeteer) to trace webpage from command line.


CHANGELOG.md

1.5.0 / 2021-01-29
==================

  * Bump lodash, acorn & https-proxy-agent
  * Add `--useragent` to set user agent

1.4.3 / 2019-09-11
==================

  * Setup and enforce eslint-config-makina as ESLint ruleset
  * Upgrade Puppeteer for security concerns

1.4.2 / 2019-01-03
==================

  * Fix vulnerabilities through `npm audit fix`

1.4.1 / 2018-02-13
==================

  * Remove extraneous space after `--timeout` flag definition
  * Return 1 as exit code on any catched exception

1.4.0 / 2018-02-12
==================

  * Debug `--headless` switch by inverting behavior
  * Offer a new flag for setting up default timeout duration
  * Avoid asking Puppeteer to clip element on fullpage capture
  * Catch errors from any calls to Puppeteer to finally close browser

1.3.1 / 2018-02-12
==================

  * Add cross reference to puppeteer-trace-cli
  * Remove an irregular whitespace
  * Update readme file to match current `--help` content

1.3.0 / 2018-02-11
==================

  * Improve main args processing
  * Unhide `--headless` option in `--help`
  * Use help texts from Puppeteer
  * Remove default type & quality settings
  * Improve `--help` and corresponding texts in readme

1.2.0 / 2018-02-10
==================

  * Redirect picture stream to stdout when `--output -` or `-o -`
  * Redirect picture stream to stdout when no `--output` given
  * Define minimum required node version to 6.4.0

1.1.1 / 2018-02-09
==================

  * Create command for updating changelog when releasing new version
  * Indicate default viewport in readme file
  * Change default suggested install source from Github to npm
  * Define github repository in package.json
  * Create changelog and rename main readme

1.1.0 / 2018-02-09
===================

  * Add eslint rules
  * Use STDout insead of console log & define exit codes
  * Use a constant for EOL char
  * Define language for readme code block

1.0.0 / 2018-02-09
===================

  * Add keywords, author, and signature
  * Enable command line access from package install
  * Create minimalist Readme
  * Ignore node_modules directory & default screenshot files



index.js
#!/usr/bin/env node

const puppeteer = require('puppeteer');

const EOL       = '\n';

const argsDef = [
  /* eslint-disable no-multi-spaces */
  { name: 'url',      alias: 'u', type: String,  description: `URL to navigate page to. The url should include scheme, e.g. https://.${EOL}`, defaultOption: true },
  { name: 'output',   alias: 'o', type: String,  description: `The file path to save the image to. If path is a relative path, then it is resolved relative to current working directory. If no path is provided, the image won't be saved to the disk.${EOL}` },
  { name: 'selector', alias: 's', type: String,  description: `A CSS selector of an element to wait for. \n[italic]{Default: body}${EOL}` },
  { name: 'type',     alias: 't', type: String,  description: `Specify screenshot type, can be either jpeg or png. \n[italic]{Default: png}${EOL}` },
  { name: 'quality',  alias: 'q', type: Number,  description: `The quality of the image, between 0-100. Not applicable to png images.${EOL}` },
  { name: 'width',    alias: 'w', type: Number,  description: `Viewport width in pixels \n[italic]{Default: 800}${EOL}` },
  { name: 'height',   alias: 'h', type: Number,  description: `Viewport height in pixels \n[italic]{Default: 600}${EOL}` },
  { name: 'useragent',            type: String,  description: `The user agent of the browser \n[italic]{Default: undefined}${EOL}` },
  { name: 'timeout',              type: Number,  description: `Maximum time to wait for in milliseconds. \n[italic]{Default: 30000}${EOL}` },
  { name: 'fullPage', alias: 'f', type: Boolean, description: `When true, takes a screenshot of the full scrollable page. \n[italic]{Defaults: false}.${EOL}` },
  { name: 'noheadless',           type: Boolean, description: `Allow disabling headless mode. \n[italic]{Default: false}${EOL}` },
  { name: 'help',     alias: '?', type: Boolean, description: `This help${EOL}` },
  /* eslint-enable no-multi-spaces */
];

const args  = require('command-line-args')(argsDef);
const usage = require('command-line-usage')({ header: 'Headless screenshot with Puppeteer', optionList: argsDef });

const doCapture = async function doCapture ({
  url,
  output,
  type,
  quality,
  noheadless,
  selector = 'body',
  width = 800,
  height = 600,
  useragent = undefined,
  timeout = 30000,
  fullPage = false,
}) {
  const browser = await puppeteer.launch({ headless: !noheadless });
  const page    = await browser.newPage();

  page.setDefaultNavigationTimeout(timeout);

  try {
    if (useragent) {
      await page.setUserAgent(useragent);
    }

    await page.setViewport({ width, height });

    await page.goto(url, { waitUntil: ['load', 'networkidle0'] });

    await page.waitForSelector(selector, { visible: true, timeout });

    const outputPath = output === '-' ? undefined : output;

    const picture = await page.screenshot({
      type: type === 'jpg' ? 'jpeg' : type,
      quality,
      fullPage,
      path: outputPath,
      clip: !fullPage && await (await page.$(selector)).boundingBox(),
    });

    if (!outputPath) {
      process.stdout.write(picture);
    }
  } catch (error) {
    await browser.close();
    process.exitCode = 1;
    throw error;
  }

  await browser.close();
};


if (args.help || !args.url) {
  !args.help && process.stderr.write(`No url provided.${EOL}`);
  process.stderr.write(usage);
  process.exitCode = 1;
} else {
  doCapture(args);
}
